### DISCUSSION TITLE
How to make defineElement work with typescript?

### DISCUSSION THREAD

**Discussion by slaweet** [vueform/vueform#191] [Questions]
### Context
I'm trying to create a custom element used in https://wails.io/ based app, but I assume that's not relevant. The essence of the component is the snippet below. 

### Expected behaviour
This component (or a slight modification of it) works with no typescript errors
```vue
<template>
  <ElementLayout>
    <template #element>
      <button
        type="button"
        @click="selectPath"
      >
        Select file
      </button>

      <input
        :id="fieldId"
        :value="value"
        type="text"
        v-bind="{ ...aria }"
        @input="handleInput"
      >
    </template>
  </ElementLayout>
</template>


<script lang="ts">
import { PropType, toRefs } from 'vue';
import { defineElement } from '@vueform/vueform';

import { OpenFileDialog } from '@wailsjs/go/main/App';


export default defineElement({
  name: 'FileselectElement',
  props: {
    directory: Boolean,
    filters: {
      type: Array as PropType<{ DisplayName: string, Pattern: string}[] | null>,
      default: () => null,
    },
  },
  setup(props: any, { element }: any) {
    const { handleInput, value } = element;
    const { filters } = toRefs(props);

    async function selectPath(): Promise<void> {
      const result = OpenFileDialog({ DefaultFilename: value.value, ...(filters.value && { filters: filters.value }) });

      if (result) {
        element.update(result);
      }
    }

    return {
      ...element,
      selectPath,
      handleInput,
      value,
    };
  },
});
</script>
```

### Actual behaviour
The user-facing functionality works well only if I add `//@ts-nocheck` comment to the beginning of the script to skip typescript checks. But if I want typescript enabled, I get the following errors (Essentially all the stuff exposed through `setup` is not available in the template). 

```shell
vue-tsc --noEmit

src/components/FileSelectElement.vue:3:16 - error TS2339: Property 'element' does not exist on type '{ field: VNode<RendererNode, RendererElement, { [key: string]: any; }>[]; label: VNode<RendererNode, RendererElement, { ...; }>[]; ... 4 more ...; after: VNode<...>[]; }'.
3     <template #element>
                 ~~~~~~~
src/components/FileSelectElement.vue:6:17 - error TS2339: Property 'selectPath' does not exist on type '{}'.
6         @click="selectPath"
                  ~~~~~~~~~~
src/components/FileSelectElement.vue:12:14 - error TS2339: Property 'fieldId' does not exist on type '{}'.
12         :id="fieldId"
                ~~~~~~~
src/components/FileSelectElement.vue:13:17 - error TS2339: Property 'value' does not exist on type '{}'.
13         :value="value"
                   ~~~~~
src/components/FileSelectElement.vue:15:22 - error TS2339: Property 'aria' does not exist on type '{}'.
15         v-bind="{ ...aria }"
                        ~~~~
src/components/FileSelectElement.vue:16:17 - error TS2339: Property 'handleInput' does not exist on type '{}'.
16         @input="handleInput"
                   ~~~~~~~~~~~
```

Any suggestions how to make this work with typescript? Or a link to an example app where it works with typescript?

**Comment by matthiasPOE** [comment]
@adamberecz any chance you can chime in here we have the same problem that we need to add `//@ts-nocheck` everywhere which is not an option in the long run

**Comment by adamberecz** [comment]
Unfortunately Vueform is behind in TS support. It only has a generated definition file which quite simple.

Could [this](https://github.com/vueform/vueform/blob/main/types/index.d.ts#L180) be something useful here?

**Reply by matthiasPOE** [reply]
Hi @adamberecz thanks yea unfortunately its not useable. In our case we are also using Vue Composition API so the declarations are completely missing when creating custom form builder elements

**Reply by adamberecz** [reply]
Vueform 1.0 wasn’t built with TypeScript in mind, and that’s the core issue here. We might be able to find a _semi-hacky_ workaround to make it function without using `//@ts-nocheck`, but that wouldn’t make it truly type-safe. The next version will have full TypeScript support, so in the long run, that’s the real solution to this problem.

**Reply by matthiasPOE** [reply]
Awesome stuff thats great to hear!

**Comment by BaptisteCrouzet** [comment]
IMO the use of `//@ts-nocheck` is a bad solution. But ok, read it, a version 2 seems to be planned.
Is there any roadmap or an estimation of the possible release date ?

Any way to help on this work ?

**Reply by adamberecz** [reply]
I'm working on an official announcement of 2.0 – the first public release can be expected around Q2.

Appreciate the offer –  In the meantime fixing issues in [types/index.d.ts](https://github.com/vueform/vueform/blob/main/types/index.d.ts) would be a great way of contributing and possibly making v1 type-safe as well.
